### PR TITLE
detecting IPSEC SPI unchanged

### LINKED ISSUE #14873: Throw error if IPSec key changed but key ID didn't
The key rotation for transparent encryption should always be performed with a new key ID [as per the documentation](https://docs.cilium.io/en/latest/gettingstarted/encryption/#key-rotation). When reading the IPSec key, we should throw an error if we detect that the key changed while the key ID didn't.

The key is loaded in [`loadIPSecKeys()`](https://github.com/cilium/cilium/blob/5b911b80127d7d79120252334b92f7fd6495831c/pkg/datapath/linux/ipsec/ipsec_linux.go#L445) and we compare the key IDs at: https://github.com/cilium/cilium/blob/5b911b80127d7d79120252334b92f7fd6495831c/pkg/datapath/linux/ipsec/ipsec_linux.go#L548

- [ ] Implement the error case.
- [ ] Test it by doing a key rotation with an unchanged key ID.
- [ ] Add a unit test in [`ipsec_linux_test.go`](https://github.com/cilium/cilium/blob/5b911b80127d7d79120252334b92f7fd6495831c/pkg/datapath/linux/ipsec/ipsec_linux_test.go).

### PR DESCRIPTION
**Pull request by smagnani96**
Adding logic (and unit test) to detect unchanged SPI while creating a new IPSEC key.

Fixes: #14873

@pchaigno @julianwiedmann :smiley: 

```release-note
In case of an IPsec key rotation, error if the user forgot to increment the SPI per the documentation.
```

### CODE REVIEW

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (rgo3) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go]: Nit: can't we reuse the `keys` variable?

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (rgo3) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go]: This test expects behavior I'm a bit worried about: Are we sure that LoadIPSecKeys will never be called on the same key file in practice? Or that the key file has two consecutive entries that reference the same key and SPI? From the original issue I understand that we should only fail if the key changed but SPI didn't. The introduced behavior doesn't exactly match that IIUC.

**Review by rgo3 — CHANGES_REQUESTED** [review]
When I was assigned to the issue that's being addressed here, the originally proposed solution involved checking the key and SPI from an existing XFRM state and comparing it to the key and SPI we read in `LoadIPSecKeys`. Overall I prefer the simplicity of the approach here but `LoadIPSecKeys` not behaving the same if it's called on the same key file even if underlying kernel state has not changed worries me a bit.

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (smagnani96) [reply pkg/datapath/linux/ipsec/ipsec_linux_test.go]: I see, thanks for pointing out :smiley: 
The new code should catch that behavior.

**Comment by smagnani96** [comment]
> When I was assigned to the issue that's being addressed here, the originally proposed solution involved checking the key and SPI from an existing XFRM state and comparing it to the key and SPI we read in `LoadIPSecKeys`. Overall I prefer the simplicity of the approach here but `LoadIPSecKeys` not behaving the same if it's called on the same key file even if underlying kernel state has not changed worries me a bit.

I modified the code to:

1. allow providing two identical keys. Though, now the `ipSecKeysRemovalTime[oldKey.Spi] = time.Now()` is executed only when needed;
2. throw error when different keys with same spi.
3. changed the tests accordingly. In particular, the SPI of all the keys variables declared for the legit `TestLoadKeys`, since now different keys with same SPI is not allowed.

My last doubt is: why would someone have more than 1 key (duplicate or even different ones) in the secret file?  Is it for backward compatibility?

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go:43]: Doesn't changing this test case to have a single line also reduces the scope of existing tests?

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go:92]: Isn't there an implicit dependency between these two test cases? If someone removes the first, or loads a new key in between, it will break the second even though the second won't have been modified. Or did I miss something?

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go]: Should we keep this exactly the same as `keysSameSpiDat`, just with different SPI and key? If the goal is to check that we mandate a new SPI, then let's test that and nothing else.

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go]: It may be worth checking that we do fail with the expected error here. If the test starts failing with something else, it might hide a regression.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Moving this line inside the if may change the behavior since it will now be skipped in case the key didn't change. Is there any wider impact?

Comment (smagnani96) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: I made a different PR discussing it https://github.com/cilium/cilium/pull/34652.
At this point, here we introduce just the check for the same SPI, for which we want to return an error. In this case, I would not insert the `oldKey.Spi`  in the `ipSecKeysRemovalTime` map, since it is not actually removed.
What's your opinion on this?

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (smagnani96) [reply pkg/datapath/linux/ipsec/ipsec_linux_test.go:43]: Here now I'm leaving the variables as they are, just changing the SPI in each of them to make sure the existing tests don't fail because of this reason. In addition, I'm introducing `keysSameSpiDat` which defines two valid keys but with the same SPI: this is used to cover the new case introduced.

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (smagnani96) [reply pkg/datapath/linux/ipsec/ipsec_linux_test.go:92]: Right now I opted for introducing and leaving only the `TestLoadKeysSameSPI` test, which tests the actual use case when a new valid key has been provided without increasing the SPI. I think this test should be enough to logically test what is being introduced and addressed with this PR.

**Comment by smagnani96** [comment]
This PRs has been rebased based on https://github.com/cilium/cilium/pull/34652 (1st commit 4e74c4a08330a6f41a0b3580171ff5c5425b8b0e).

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: Yep, it looks good now.

**Review by pchaigno — APPROVED** [review]
I think this is good to go without the first commit.

I'd remove it so we can merge because I suspect the other PR will take a bit longer to get ready.

**Review by pchaigno — APPROVED** [review]

File: pkg/datapath/linux/node_linux_test.go
Comment (pchaigno) [inline pkg/datapath/linux/node_linux_test.go]: Is this really necessary? This sort of global variable in tests can easily create side effects that then end up introducing flakes.

Comment (smagnani96) [reply pkg/datapath/linux/node_linux_test.go]: AFAIU, in these tests we just need an IPSEC key to be present, and we never remove it between these tests.
Possible solutions:

1. Let the tests call the `LoadIPSecKey` with an incremental SPI each time.
2. Let the tests call the `LoadIPSecKey` only once.

Either way, instead of global var, we could wrap the incremental SPI or a boolean variable. Would this be a preferred way? 

```go
var loadOnceIPSecKey func(t *testing.T, log *slog.Logger) = func() func(t *testing.T, log *slog.Logger) {
	var loaded bool
	inner := func(t *testing.T, log *slog.Logger) {
		if loaded {
			return
		}
		keys := bytes.NewReader([]byte("6 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f1 128\n"))
		_, _, err := ipsec.LoadIPSecKeys(log, keys)
		require.NoError(t, err)
		loaded = true
	}
	return inner
}()
```

Comment (pchaigno) [reply pkg/datapath/linux/node_linux_test.go]: Could we clean up the key after each test? That way we avoid state leaking from one test to the next. Then it doesn't matter what each test does because they're all independent.

Comment (smagnani96) [reply pkg/datapath/linux/node_linux_test.go]: I'm scared we don't have a method yet to clean up the key held in `ipSecKeysGlobal`, unless providing a new key with different SPI. Am I missing something? In case, would it be worth adding? 
Otherwise, for the just sake of this test a `sync.Once` could be enough, but I do recognize that what you're suggesting would be a way more elegant solution.

Comment (pchaigno) [reply pkg/datapath/linux/node_linux_test.go]: I'm just suggesting to do the cleanup from `setupIPSecSuitePrivileged` after each test instead of once at the end. Am I missing something?

Comment (smagnani96) [reply pkg/datapath/linux/node_linux_test.go]: I think I got what you meant, hopefully :smiley:

**Comment by jibi** [comment]
removing myself as Paul has already approved on behalf of sig-datapath

File: pkg/datapath/linux/node_linux_test.go
Comment (rgo3) [reply pkg/datapath/linux/node_linux_test.go]: IIUC, shouldn't the new API you added, `UnsetTestIPSecKey()`, be called from `setupIPSecSuitePrivileged` as well? Now it's only called in the node test suite.

**Review by rgo3 — APPROVED** [review]
Left one comment with regards to cleaning state after individual tests, otherwise LGTM. Sorry for the long delay on my part.

File: pkg/datapath/linux/node_linux_test.go
Comment (smagnani96) [reply pkg/datapath/linux/node_linux_test.go]: > IIUC, shouldn't the new API you added, UnsetTestIPSecKey(), be called from setupIPSecSuitePrivileged as well? Now it's only called in the node test suite.

Since `setupIPSecSuitePrivileged` is within the same package, in theory it wouldn't be necessary to invoke the public API (so far it directly reinitialize the `ipSecKeysGlobal`). What would you suggest?

Comment (rgo3) [reply pkg/datapath/linux/node_linux_test.go]: I don't have a strong opinion on this but since we have a public API that is needed outside the package I don't see why we shouldn't use it inside the package itself to make the cleanup of the key consistent in and outside of the ipsec package?

Comment (smagnani96) [reply pkg/datapath/linux/node_linux_test.go]: I see your point, good to have coherent behavior especially while tearing down test suites. Modified the latest commit, let me know :smiley:

Comment (rgo3) [reply pkg/datapath/linux/node_linux_test.go]: LGTM. I'll remove the discussion label so we can get the PR merged. If Paul still sees something that is missing we can address it in a follow up I presume.

**Comment by julianwiedmann** [comment]
@smagnani96 fyi there's an outstanding discussion here, once that is resolved the PR should auto-merge.

**Comment by smagnani96** [comment]
> @smagnani96 fyi there's an outstanding discussion here, once that is resolved the PR should auto-merge.

Yep, I wanted to wait for the last feedback from @pchaigno to make sure I got what he meant :+1:

**Pull request merged**
